File: html2yaml.rb
#!/usr/bin/env ruby

########################################################################
# html2yaml.rb: HTML to YAML Converter Script
#
# Description:
# This script parses an HTML document and converts it into a structured
# YAML format. It uses Nokogiri for HTML parsing and provides options for
# including or excluding specific elements based on CSS or XPath selectors.
# This is useful for converting HTML structures into a format that can be
# easily processed by scripts or programs.
#
# Author: id774 (More info: http://id774.net)
# Source Code: https://github.com/id774/scripts
# License: LGPLv3 (Details: https://www.gnu.org/licenses/lgpl-3.0.html)
# Contact: idnanashi@gmail.com
#
# Usage:
# Run the script with the URL or path of the HTML file, along with optional
# arguments for inclusion or exclusion of elements:
#   html2yaml.rb [options] URL_OR_PATH
#
# Options:
#   -i, --include=XPATH_OR_CSS  Include elements matching the given selector
#   -x, --exclude=XPATH_OR_CSS  Exclude elements matching the given selector
#
# Example:
#   html2yaml.rb --exclude='//div[@class="footer"]' http://example.com
#   This command will parse the HTML from example.com, excluding divs with class "footer".
#
# Input/Output Example:
# - Given an HTML document like:
#   <!DOCTYPE html>
#   <html>
#   <head>
#       <title>Example Page</title>
#   </head>
#   <body>
#       <div id="main" class="container">
#           <h1>Welcome to the Example Page</h1>
#           <p>This is a sample paragraph.</p>
#       </div>
#   </body>
#   </html>
# - The output will be a YAML format representing the structure of the HTML:
#   :name: html
#   :children:
#     - :name: head
#       :children:
#         - :name: title
#           :children:
#             - :name: text
#               :value: Example Page
#     - :name: body
#       :children:
#         - :name: div
#           :attributes:
#             :id: main
#             :class: container
#           :children:
#             - :name: h1
#               :children:
#                 - :name: text
#                   :value: Welcome to the Example Page
#             - :name: p
#               :children:
#                 - :name: text
#                   :value: This is a sample paragraph.
#
########################################################################

require 'nokogiri'
require 'open-uri'
require 'optparse'
require 'yaml'

# Converts an HTML node to a YAML-structured data object
def pod_from_node(node)
  data = YAML::Omap.new

  # Extract node name
  data[:name] = node.node_name

  # Extract and sort attributes of the node
  attrs = YAML::Omap.new
  node.attribute_nodes.sort_by(&:node_name).each do |attr|
    attrs[attr.node_name] = attr.value
  end
  data[:attributes] = attrs if attrs.size > 0

  # Recursively process children nodes
  children = node.children.select(&:element?).map {|child|
    pod_from_node(child)
  }
  data[:children] = children if children.size > 0

  data
end

# Prunes HTML based on specified rules for inclusion or exclusion
def prune_html(html, rules)
  rules.each do |rule|
    type, selector = *rule
    html.root.search(selector).each do |node|
      # Mark nodes for exclusion or inclusion
      if type == :exclude
        node['html2yaml-exclude'] = 'yes'
      else
        node['html2yaml-exclude'] = 'no'
        node.ancestors.each { |ancestor| ancestor['html2yaml-exclude'] = 'no' }
      end
    end
  end

  # Remove excluded nodes and clean up attributes
  html.root.search('[html2yaml-exclude = yes]').remove
  html.root.traverse { |node| node.remove_attribute('html2yaml-exclude') }

  html
end

def main
  rules = []
  opt = OptionParser.new
  # Define command-line options for inclusion and exclusion
  opt.on('-i', '--include=XPATH_OR_CSS') { |v| rules << [:include, v] }
  opt.on('-x', '--exclude=XPATH_OR_CSS') { |v| rules << [:exclude, v] }
  opt.parse!(ARGV)

  # Load the HTML source
  source = ARGV.shift or abort "Usage: html2yaml.rb URL_OR_PATH"
  html = Nokogiri::HTML(open(source))
  html = prune_html(html, rules)

  # Convert HTML to YAML and output
  data = pod_from_node(html.root)
  puts YAML.dump(data)
end

main

File: html2yaml.py
#!/usr/bin/env python

########################################################################
# html2yaml.py: HTML to YAML Converter Script
#
#  Description:
#  This script parses an HTML document and converts it into a structured
#  YAML format. It uses BeautifulSoup for HTML parsing and PyYAML for
#  YAML generation. The script provides options for including or excluding
#  specific elements based on CSS or XPath selectors. This is useful for
#  converting HTML structures into a format that can be easily processed
#  by scripts or programs. Now, it checks for the necessary libraries
#  and exits with an error message if they are not installed.
#
#  Author: id774 (More info: http://id774.net)
#  Source Code: https://github.com/id774/scripts
#  License: The GPL version 3, or LGPL version 3 (Dual License).
#  Contact: idnanashi@gmail.com
#
#  Version History:
#  v1.5 2025-07-01
#       Standardized termination behavior for consistent script execution.
#  v1.4 2025-06-23
#       Unified usage output to display full script header and support common help/version options.
#  v1.3 2025-04-14
#       Unify error and info message formatting with stderr and prefix tags.
#  v1.2 2024-01-11
#       Added checks for BeautifulSoup and PyYAML libraries.
#  v1.1 2023-12-08
#       Removed f-strings for compatibility with Python versions below 3.6.
#  v1.0 2023-12-07
#       Initial release.
#
#  Dependencies:
#  - BeautifulSoup
#  - PyYAML
#  - requests
#
#  Notes:
#  This script requires the following Python libraries: BeautifulSoup,
#  PyYAML, and requests. These can be installed via pip using:
#    pip install bs4 pyyaml requests
#
#  Usage:
#  Run the script with the URL or path of the HTML file, along with optional
#  arguments for inclusion or exclusion of elements:
#    html2yaml.py [options] URL_OR_PATH
#
#  Options:
#    -i, --include=XPATH_OR_CSS  Include elements matching the given selector
#    -x, --exclude=XPATH_OR_CSS  Exclude elements matching the given selector
#
#  Example:
#    html2yaml.py --exclude='//div[@class="footer"]' http://example.com
#
########################################################################

import os
import sys

# Importing necessary libraries
try:
    import requests
    import yaml
    from bs4 import BeautifulSoup
except ImportError as e:
    libraries_installed = False
else:
    libraries_installed = True

def usage():
    """ Display the script header as usage information and exit. """
    script_path = os.path.abspath(__file__)
    in_header = False
    try:
        with open(script_path, 'r', encoding='utf-8') as f:
            for line in f:
                if line.strip().startswith('#' * 10):
                    if not in_header:
                        in_header = True
                        continue
                    else:
                        break
                if in_header and line.startswith('#'):
                    if line.startswith('# '):
                        print(line[2:], end='')
                    else:
                        print(line[1:], end='')
    except Exception as e:
        print("Error reading usage information: %s" % str(e), file=sys.stderr)
        sys.exit(1)
    sys.exit(0)

def html_to_yaml(element):
    """ Convert HTML element to YAML-structured data. """
    data = {'name': element.name}

    if element.attrs:
        data['attributes'] = element.attrs

    children = [html_to_yaml(child)
                for child in element.children if child.name]
    if children:
        data['children'] = children

    return data

def main():
    if not libraries_installed:
        print("[ERROR] Required libraries not installed.", file=sys.stderr)
        sys.exit(1)

    source = sys.argv[1]

    try:
        if source.startswith('http://') or source.startswith('https://'):
            response = requests.get(source)
            html = BeautifulSoup(response.text, 'html.parser')
        else:
            with open(source, 'r') as file:
                html = BeautifulSoup(file, 'html.parser')

        yaml_data = html_to_yaml(html)
        print(yaml.dump(yaml_data))
        return 0
    except Exception as e:
        print("[ERROR] Error processing HTML: {}".format(e), file=sys.stderr)
        return 2


if __name__ == "__main__":
    if len(sys.argv) < 2 or sys.argv[1] in ('-h', '--help', '-v', '--version'):
        usage()
    sys.exit(main())
